src/ui/marker.js

```diff
@@ -500,20 +519,34 @@ export default class Marker extends Evented {
     }
 
     _calculateRotation(): number {
-        if (this._rotationAlignment === "viewport" || this._rotationAlignment === "auto") {
-            return this._rotation;
-        } if (this._map && this._rotationAlignment === "map") {
-            const pos = this._pos;
-            const map = this._map;
-            if (pos && map && map._usingGlobe()) {
+        const pos = this._pos;
+        const map = this._map;
+        if (!map || !pos) { return 0; }
+        const alignment = this.getRotationAlignment();
+        if (alignment === 'map') {
+            if (map._showingGlobe()) {
                 const north = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat + .001));
                 const south = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat - .001));
                 const diff = south.sub(north);
                 return this._rotation + radToDeg(Math.atan2(diff.y, diff.x)) - 90;
             }
-            return this._rotation - this._map.getBearing();
+            return this._rotation - map.getBearing();
+        } else if (this._rotationAlignment === "horizon" && map._showingGlobe()) {
+            const zoom = map.getZoom();
+            const centerPoint = globeCenterToScreenPoint(map.transform);
+            let zoomTransition = 1;
+            if (zoom > ALIGN_TO_HORIZON_BELOW_ZOOM) {
+                const smooth = smoothstep(ALIGN_TO_HORIZON_BELOW_ZOOM, ALIGN_TO_SCREEN_ABOVE_ZOOM, zoom);
+                centerPoint.y += smooth * map.transform.height;
+                zoomTransition = 1 - smooth;
+            }
+            const rel = pos.sub(centerPoint);
+            const angle = radToDeg(Math.atan2(rel.y, rel.x));
+            const up = angle > 90 ? angle - 270 : angle + 90;
+            return up * zoomTransition + this._rotation;
+
         }
-        return 0;
+        return this._rotation;
```

Comment:
I think this can be simplified a bit (untested so please only take that as a starting point!):

- Reduce number of return
- Store the result in an offset variable, since the result can be folded as `rotation + offset`, also make the logic easier to reason about (the default behavior is _no offset_, having an anchor _applies an offset_ to the initial rotation)
- Remove an extra `if`, since the `smoothstep` is already returning 0 for values below `ALIGN_TO_HORIZON_BELOW_ZOOM` and its evaluation is trivial enough that it's not worth adding an extra branch
```js
let offset = 0;
if (alignment === 'map') {
    if (map._showingGlobe()) {
        const north = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat + .001));
        const south = map.project(new LngLat(this._lngLat.lng, this._lngLat.lat - .001));
        const diff = south.sub(north);
        offset = radToDeg(Math.atan2(diff.y, diff.x)) - 90;
    } else {
        offset = -map.getBearing();
    }
} else if (this._rotationAlignment === "horizon" && map._showingGlobe()) {
    const centerPoint = globeCenterToScreenPoint(map.transform);
    const smooth = smoothstep(ALIGN_TO_HORIZON_BELOW_ZOOM, ALIGN_TO_SCREEN_ABOVE_ZOOM, map.getZoom());
    centerPoint.y += smooth * map.transform.height;
    const zoomTransition = 1 - smooth;
    const rel = pos.sub(centerPoint);
    const angle = radToDeg(Math.atan2(rel.y, rel.x));
    const up = angle > 90 ? angle - 270 : angle + 90;
    offset = up * zoomTransition;
}
return this._rotation + offset;
```

Reply:
Good ideas Karim! I've implemented those changes and also combined the function into `_calculateZTransform()`